src/helpers/form.ts

```diff
@@ -0,0 +1,37 @@
+import { emailRegex } from "./regex"
+
+export type SignupErrorCodeType = "auth/email-already-in-use" | "auth/invalid-email" | "auth/weak-password"
+export type SigninErrorCodeType = "auth/invalid-email" | "auth/user-disabled" | "auth/user-not-found" | "auth/wrong-password"
+type errorCodeType = SignupErrorCodeType | SigninErrorCodeType
+
+export type SignupErrorType = {
+  code: SignupErrorCodeType
+} | undefined
+
+export type SigninErrorType = {
+  code: SigninErrorCodeType
+} | undefined
+
+type errorType = SignupErrorType | SigninErrorType | null
+
+// export const validateEmail = (email: string) => email.match(emailRegex)
+export const validateEmail = (email: string, callback: (arg0: any) => void, code: SignupErrorCodeType = "auth/invalid-email") => {
+  const validEmail = email.match(emailRegex)
+
+  if (typeof callback === 'function') {
+    callback(!validEmail ? {
+      code
+    } : null)
+  }
+}
+
+// TODO: remove 'any' type here ...
+export const validatePassword = (password: string, callback: (arg0?: any) => void, code: errorCodeType = "auth/weak-password") => {
+  const validPassword = password.length >= 6
+
+  if (typeof callback === 'function') {
+    callback(!validPassword ? {
+      code
+    } : null)
+  }
+}
```

Comment:
With the help of Andrew, we determined what the typing error was. Basically in `validateEmail` the `callBack` was either of type `SignupErrorType | SigninErrorType`. If we passed in `setState<SignupErrorType>`, and no `code`, `code` is by default type 'SigninErrorCodeType'. the `setState` that was passed in cannot take  'SigninErrorCodeType'.

I think we should simplify these codes and NOT have callbacks. And just have these function return true of false if valid, and then handle it with a function directly in line. The benefits of doing that would make debugging easier AND we are handing the setState explicitly in the component, which again is easier to debug